Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean

```diff
@@ -0,0 +1,205 @@
+/-
+Copyright (c) 2023 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Localization.Opposite
+
+/-!
+# Calculus of fractions
+
+Following the definitions by [Gabriel and Zisman][gabriel-zisman-1967],
+given a morphism property `W : MorphismProperty C` on a category `C`,
+we introduce the class `W.HasLeftCalculusOfFractions`. The main
+result (TODO) is that if `L : C ⥤ D` is a localization functor for `W`,
+then for any morphism `L.obj X ⟶ L.obj Y` in `D`, there exists an auxiliary
+object `Y' : C` and morphisms `g : X ⟶ Y'` and `s : Y ⟶ Y'`, with `W s`, such
+that the given morphism is a sort of fraction `g / s`, or more precisely of
+the form `L.map g ≫ (Localization.isoOfHom L W s hs).inv`.
+
+## References
+
+* [P. Gabriel, M. Zisman, *Calculus of fractions and homotopy theory*][gabriel-zisman-1967]
+
+-/
+
+namespace CategoryTheory
+
+open Category
+
+namespace Functor
+
+lemma congr_map_conjugate {C D : Type _} [Category C] [Category D] {F₁ F₂ : C ⥤ D}
+    (h : F₁ = F₂) {X Y : C} (f : X ⟶ Y) :
+    F₁.map f = eqToHom (by congr) ≫ F₂.map f ≫ eqToHom (by symm; congr) := by
+  subst h
+  simp
+
+end Functor
+
+namespace MorphismProperty
+
+variable {C D : Type _} [Category C] [Category D]
+
+/-- A left fraction from `X : C` to `Y : C` for `W : MorphismProperty C` consists of the
+datum of an object `Y' : C` and maps `f : X ⟶ Y'` and `s : Y ⟶ Y'` such that `W s`. -/
+structure LeftFraction (W : MorphismProperty C) (X Y : C) where
+  /-- the auxiliary object of a left fraction -/
+  {Y' : C}
+  /-- the numerator of a left fraction -/
+  f : X ⟶ Y'
+  /-- the denominator of a left fraction -/
+  s : Y ⟶ Y'
+  /-- the condition that the denominator belongs to the given morphism property -/
+  hs : W s
+
+namespace LeftFraction
+
+variable (W : MorphismProperty C) {X Y : C}
+
+/-- The left fraction from `X` to `Y` given by a morphism `f : X ⟶ Y`. -/
+@[simps]
+def ofHom (f : X ⟶ Y) [W.ContainsIdentities] :
+    W.LeftFraction X Y := mk f (𝟙 Y) (W.id_mem Y)
+
+variable {W}
+
+/-- The left fraction from `X` to `Y` given by a morphism `s : Y ⟶ X` such that `W s`. -/
+@[simps]
+def ofInv (s : Y ⟶ X) (hs : W s) :
+    W.LeftFraction X Y := mk (𝟙 X) s hs
+
+/-- If `φ : W.LeftFraction X Y` and `L` is a functor which inverts `W`, this is the
+induced morphism `L.obj X ⟶ L.obj Y`  -/
+noncomputable def map (φ : W.LeftFraction X Y) (L : C ⥤ D) (hL : W.IsInvertedBy L) :
+    L.obj X ⟶ L.obj Y :=
+  have := hL _ φ.hs
+  L.map φ.f ≫ inv (L.map φ.s)
+
+@[reassoc (attr := simp, nolint unusedHavesSuffices)]
```

Comment:
This nolint points at a bug. Is there already an issue for that?

Reply:
Using `letI` instead of `have` fixed the problem...